src/vm/callables.rs

```diff
@@ -59,15 +67,30 @@ impl DefinedFunction {
             Err(CheckErrors::IncorrectArgumentCount(self.arguments.len(), args.len()))?
         }
 
-        let arg_iterator = self.arguments.iter().zip(self.arg_types.iter()).zip(args.iter());
-        for ((arg, type_sig), value) in arg_iterator {
-            if !type_sig.admits(value) {
-                return Err(CheckErrors::TypeValueError(type_sig.clone(), value.clone()).into())
-            }
-            if let Some(_) = context.variables.insert(arg.clone(), value.clone()) {
-                return Err(CheckErrors::NameAlreadyUsed(arg.to_string()).into())
+        let mut arg_iterator: Vec<_> = self.arguments.iter().zip(self.arg_types.iter()).zip(args.iter()).collect();
+
+        for arg in arg_iterator.drain(..) {
+            let ((name, type_sig), value) = arg;
+            match (type_sig, value) {
+                (TypeSignature::TraitReferenceType(trait_reference), Value::Principal(PrincipalData::Contract(contract_id))) => {
+                    // Argument is a trait reference, probably leading to a dynamic contract call
+                    // This is the moment when we're making sure that the target contract is 
+                    // conform.
+                    let trait_identifier = env.contract_context.lookup_trait_reference(trait_reference)
```

Comment:
If the type-checker is working correctly, then this should never fail with `TraitReferenceUnknown`, right?  That would imply that it was possible to instantiate contracts that call non-existent functions, right?

Reply:
Correct, this should never fail, but I need to unwrap :).
You think we should panic instead of throwing?

Reply:
If you're certain that this will never fail, then it should probably `unwrap()` or `expect()` (but let's confirm with @kantai).

Reply:
For errors that the static analysis should catch, the convention we've used to just throw the CheckError here. That makes testing the check much easier from the unit tests (catching panics isn't as reliable for the test), but it will also allow us to do some testing in the future on a constraint that a type checked contract can never yield a CheckError type.